docs/src/main/asciidoc/building-my-first-extension-guide.adoc

```diff
@@ -0,0 +1,272 @@
+
+Building my first Quarkus Extension
+-----------------------------------
+
+https://quarkus.io/extensions/[Quarkus extensions] enhance your application just as projects dependencies do. The role of the extensions is to do all the technical piping to make third-party libraries/technologies Quarkus-compatible. This is how you can use your battle-tested ecosystem and take advantage of Quarkus perfomance and native compilation
+
+Prerequisites
+~~~~~~~~~~~~~
+
+To complete this guide, you need:
+
+* less than 30 minutes
+
+* https://github.com/quarkusio/quarkus[Quarkus sources ]
+
+* an IDE
+
+* JDK 1.8+ installed with JAVA_HOME configured appropriately
+
+* Apache Maven 3.5.3+
+
+Basic Concepts
+~~~~~~~~~~~~~~
+
+First thing first, we will need to start some basic concepts.
+
+* JVM mode Vs Native mode
+  ** Quarkus is before all a Java framework, that means you can develop, package and run classic JAR application, that's what we call *JVM mode*.
+  ** Thanks to https://graalvm.org/[GraalVM] you can compile your Java application into machine specific code (like you do in Go or C++) and that's what we call *Native Mode*.
+
+The operation of compiling Java bytecode into a native system-specific machine code is named *Ahead of Time Compilation* (aka. AoT).
+
+* Build time Vs. Run time
+  ** The Build time corresponds to all the actions you apply to your Java source files to convert them into something runnable (class files, jar/war, native images). Usually this stage is composed by the compilation, annotation processing, bytecode generation, etc. At this point, everything is under developer's scope and control.
+  ** The Run time is all the actions that happens when you execute your application. It's obviously focused on starting your business-oriented actions but it relies on a lot of technical actions like loading libraries and configuration files, scanning the application's classpath, configuring the dependency injection, setting up your Object-Relationnal Mapping, instanciating your REST controllers, etc.
+
+Usually, Java frameworks do their bootstrapping during the Run time before actually starting your application "Business oriented layer". During the bootstrap frameworks dynamically collect metadata by scanning the classpath to find configurations, entity definitions, dependency injection binding, etc. in order to instanciate through reflection the proper obejcts. The main consequences are:
+
+* Delaying the readiness of your application: you need to wait couple of seconds before actually serve a business request.
+
+* Having a peak of resource consumption at bootstrap: In a constrained environment, you will need to size the needed resources based on your technical boostrap need rather than your actual business needs.
```

Comment:
s/boostrap/bootstrap/
I'm not totally sure this is really a thing? But I must admit I never worked in contrained environments.

Reply:
I strongly agree with consequence given by @danielpetisme. For example, when on AWS EC2, a service can often work on lower resources level (CPU or IOPS), but bootstrap takes forever. I can recall case where service uptime was very important but bootstrap of all services executed simultaneously took 10 minutes. However, after service was up, the resources consumption was extremely low. So business decision was to use more resources just to boot the whole box faster.

Reply:
` During the bootstrap frameworks` would probably be better of as ` During bootstrap, frameworks`